src/Control/Carrier/NonDet/Church.hs

```diff
@@ -109,33 +108,5 @@ instance MonadTrans NonDetC where
 instance (Algebra sig m, Effect sig) => Algebra (NonDet :+: sig) (NonDetC m) where
   eff (L (L Empty))      = empty
   eff (L (R (Choose k))) = k True <|> k False
-  eff (R other)          = NonDetC $ \ fork leaf nil -> eff (handle (Leaf ()) (fmap join . traverse runNonDetA) other) >>= fold fork leaf nil
+  eff (R other)          = NonDetC $ \ fork leaf nil -> eff (handle (pure ()) (runNonDet (liftA2 (<|>)) runNonDetA (pure empty)) other) >>= runNonDet fork leaf nil
```

Comment:
It turns out there are (at least) two wrong ways and two right ways to do this.

**Wrong:**

1. using `pure . join` as the handler
2. using `fmap join . runNonDetA` as the handler

These are both wrong in that they screw up higher-order effects like `Catch` in `NonDetC (ErrorC e PureC)`—but not `ErrorC e (NonDetC PureC)`, curiously enough.

**Right:**

1. using `NonDetC PureC`
2. this, which is significantly less code than `NonDetC PureC` entailed